Fix issue in tf.image.extract_glimpse

yongtang opened the pull request:
This PR is to re-apply PR #12829. While #12829 was merged before, it was reverted at one point. Not sue about the reason though my guess is that there are some internal testing that caused the revert. This PR will try to submit again, and will fix any internal tests that fails.

This PR fixes #38545.

This fix tries to fix the issue raised in #2134 where
`tf.image.extract_glimpse` does not work as expected
when `centered=False` and `normalized=False`

This fix fixes #2134.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix issue in tf.image.extract_glimpse

This PR is to re-apply PR 12829. While 12829 was merged before,
for some reason it was reverted at one point. The guess is that
there are some internal testing that caused the revert. This
PR will try to submit again, and fix any internal tests that fails.

This fix tries to fix the issue raised in 2134 where
`tf.image.extract_glimpse` does not work as expected
when `centered=False` and `normalized=False`

This fix fixes 2134.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test cases for tf.image.extract_glimpse

Add test cases for tf.image.extract_glimpse with
centered=False and normalized=False

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix test failure caused by API changes in tests

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix test failure due to changes of the fix for centered=False and normalized=False

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix incorrect doc test

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix incorrect doc example with centered=False

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

karmel commented:
For tensorflow/api-owners

Is it the case that these examples/tests were just wrong before? We are concerned that these changes are not backwards compatible if even our examples have to change. Since this is a significant change in behavior, we should have a v2 version of the op rather than a change to the existing op.

That said, it's unclear based on https://github.com/sb2nov/tensorflow/commit/63b599bcd5443366e0f6c65bc6a349d3da25c5a4 whether this is universally considered a bug versus expected behavior. Maybe what we need here is a clarification of behavior in docstrings?

alextp commented:
I prefer option (3), coupled with a deprecation of tf.image.extract_glimpse
with instructions in the message for how to get the "broken" results using
extract_glimpse_v2.

Any other option can change the properties of code which existing works now
and depends on the buggy behavior, and because TF promises backwards
compatibility of its python API and graphdef we cannot do that.

On Tue, Apr 21, 2020 at 8:15 AM Yong Tang <notifications@github.com> wrote:

> Thanks @karmel <https://github.com/karmel> @tensorflow/api-owners
> <https://github.com/orgs/tensorflow/teams/api-owners> for the comment.
> The previous behavior of tf.image.extract_glimpse was incorrectly
> implemented in two issues:
>
>    1. both centered=True and centered=False will behave like
>    centered=False (sort of, see combination effect of 2)
>    2. there is also another issue where the centered=False will take
>    positive value incorrectly.
>
> Below is an example:
>
> import tensorflow as tf
> import numpy as np
>
> img = tf.constant(np.arange(25).reshape((1, 5, 5, 1)), dtype=tf.float32)
>
> # Image:
> # [  0.   1.   2.   3.   4.]
> # [  5.   6.   7.   8.   9.]
> # [ 10.  11.  12.  13.  14.]
> # [ 15.  16.  17.  18.  19.]
> # [ 20.  21.  22.  23.  24.]
>
> tf.image.extract_glimpse(
>     img, [3, 3], [[-2, 2]], centered=False, normalized=False, noise='zero')
>
> # [ 0.  0.  0.]
> # [ 0.  0.  0.]
> # [ 0.  0.  0.]
>
> In the above example, since size = [3, 3], offset = [[-2, 2]], and
> centered=False, the obtained region should take upper-left corner as [0,
> 0], thus the correct one should be:
>
> # [ 0.  0.  0.]
> # [ 0.  0.  0.]
> # [ 2.  3.  4.]
>
> Given this implementation had been in place for a really long time. I
> don't know the best way to address it without breaking the API. There might
> be several considerations:
>
>    1. We could consider this as a bug that must be fixed, so change the
>    underlying implementation directly.
>    2. We could implement the correct one in C++ with a new kernel (
>    ExtractGlipmpseV2), and leave the old C++ kernel alone. In this way,
>    when people use python API, they will be rerouted to the new kernel,
>    however, for saved models poeple will get the old result.
>    3. We could implement the correct one with a new python API
>    extract_glimpse_v2, and co-exist with both extract_glimpse and
>    extract_glimpse_v2. It might cause some confusion though, as user will
>    see two exactly APIs with different behavior.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/38549#issuecomment-617244215>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRLEBMICJOFHBTLZ5J3RNW2A7ANCNFSM4MICQYMA>
> .
>


-- 
 - Alex

martinwicke commented:
I think we should be a little more aggressive here. This is clearly a bug,
and it is also inconsistent with the documentation. We don't have to
slavishly preserve the broken behavior. I think therefore, option 2 is
best. This preserves the behavior of existing graphs, but we can fix the
behavior that is generated for new programs. Because this is a behavior
change, we have to make sure to note this in the relnotes.

yongtang committed:
Use ExtractGlimpseV2 and ExtractGlimpse to make sure C++ kernel is backward compatible

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Expand test case to cover both old kernel (ExtractGlimpse) and new kernel (ExtractGlimpseV2)

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add api_def_ExtractGlimpseV2.pbtxt

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update API golden

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

alextp reviewed (commented):
If you want to go with wicke's suggestion you still need a tf_export of the old extract_glimpse which is v1-only (so tf.compat.v1.extract_glimpse is the old behavior and tf.extract_glimpse is the new one).

yongtang committed:
Reroute tf.compat.v1.extract_glimpse to use gen_image_ops.extract_glimpse (old API)

This fix reroute tf.compat.v1.extract_glimpse to use gen_image_ops.extract_glimpse,
so that the behavior of TF 1.x remains the same.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks all for the feedback! I have updated the PR with the following changes:
- A new C++ kernel `ExtractGlimpseV2` coexist with old `ExtractGlimpse` kernel.
- `tf.compat.v1.image.extract_glimpse` is routed to `ExtractGlimpse`
- `tf.compat.v2.image.extract_glimpse` is routed to `ExtractGlimpseV2`
- Existing tests are not modified but routed to old `ExtractGlimpse` (so that it makes sure it does not break).
- Added new tests that is different from old tests (routed to new `ExtractGlimpseV2`, cover the new behavior)

Please take a look.

martinwicke commented:
Thank you! Can you add a bullet to relnotes explaining this (in the breaking changes section for the next release?

yongtang committed:
Update RELEADE.md to capture the breaking change of `tf.image.extract_glimpse`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @martinwicke! The PR has been updated with the following added in RELEASE.md:

 Breaking Changes
* `tf.image.extract_glimpse` has been updated to correctly process the case where  `centered=False` and `normalized=False`. This is a breaking change as the output is different from (incorrect) previous versions. Note this breaking change only impacts `tf.image.extract_glimpse` and `tf.compat.v2.image.extract_glimpse` API endpoints. The behavior of `tf.compat.v1.image.extract_glimpse` does not change. The behavior of exsiting C++ kernel `ExtractGlimpse` does not change as well, so saved models will not be impacted.

martinwicke reviewed (changes requested):
You export the ops to raw_ops, but you don't change the actual tf.image endpoint, right?

martinwicke commented:
Sorry, you do change the endpoint, but the docs are inconsistent.

yongtang commented:
Thanks @martinwicke. That was caused by doc example of v1 also use `tf.image.extract_glimpse`  in docstring. So the example has to be updated in order to pass the `doctest` check.

I think we can change the example  in v1 doc to `tf.compat.v1.image.extract_glimpse`, then the example will pass the `doctest` without behavior change.

Let me update the PR shortly.

yongtang committed:
Update doc example of v1 to keep old behavior with usage of tf.compat.v1.image.extract_glimpse

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@martinwicke The PR has been updated. The doc example in v1 now maintain the old behavior, though the API usage changed to `tf.compat.v1.image.extract_glimpse` so that python doctest could pass. Please take a look.

gbaned commented:
@yongtang Can you please resolve conflicts? Thanks!

yongtang commented:
@gbaned  The PR has been rebased and pushed. Please take a look.

gbaned commented:
@yongtang  Can you please resolve conflicts? Thanks!

mihaimaruseac committed:
Merge branch 'master' into 12829-extract_glimpse
